injector/cpu_stress_test.go

```diff
@@ -0,0 +1,68 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023 Datadog, Inc.
+package injector_test
+
+import (
+	"time"
+
+	"github.com/DataDog/chaos-controller/cpuset"
+	. "github.com/DataDog/chaos-controller/injector"
+	"github.com/DataDog/chaos-controller/mocks"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("CPU stress works", func() {
+	var (
+		config  Config
+		inj     Injector
+		cgroups *mocks.CGroupManagerMock
+		process *mocks.ProcessManagerMock
+		runtime *mocks.ProcessRuntimeMock
+	)
+
+	cpu1, cpu2, cpu3 := 0, 1, 2
+	threeCPUs := cpuset.NewCPUSet(cpu1, cpu2, cpu3)
+
+	BeforeEach(func() {
+		process = mocks.NewProcessManagerMock(GinkgoT())
+		runtime = mocks.NewProcessRuntimeMock(GinkgoT())
+		cgroups = mocks.NewCGroupManagerMock(GinkgoT())
+
+		config = Config{
+			Log:    log,
+			Cgroup: cgroups,
+		}
+	})
+
+	JustBeforeEach(func() {
+		inj = NewCPUStressInjector(config, 10, process, runtime)
+	})
+
+	Specify("expected calls to stress 3 cpus", func(ctx SpecContext) {
+		processID := 42
+		cpuCount := threeCPUs.Size()
+
+		cgroups.EXPECT().ReadCPUSet().Return(threeCPUs, nil).Once()
+		process.EXPECT().ProcessID().Return(processID).Once()
+		cgroups.EXPECT().Join(processID).Return(nil).Once()
+		process.EXPECT().Prioritize().Return(nil).Once()
+		runtime.EXPECT().GOMAXPROCS(cpuCount).Return(0).Once()
+
+		process.EXPECT().ThreadID().Return(1244).Once()
+		process.EXPECT().ThreadID().Return(1245).Once()
+		process.EXPECT().ThreadID().Return(1246).Once()
+
+		runtime.EXPECT().LockOSThread().Times(cpuCount)
+		runtime.EXPECT().UnlockOSThread().Times(cpuCount)
+
+		process.EXPECT().SetAffinity([]int{cpu1}).Return(nil).Once()
+		process.EXPECT().SetAffinity([]int{cpu2}).Return(nil).Once()
+		process.EXPECT().SetAffinity([]int{cpu3}).Return(nil).Once()
+
+		Expect(inj.Inject()).To(Succeed())
+		Expect(inj.Clean()).To(Succeed())
+	}, NodeTimeout(10*time.Second))
+})
```

Comment:
Is it possible to add unit tests to test error cases?

Reply:
Thanks for asking 🤗 
It uncovered a bug before it landed on PROD 💪